\documentclass{article}
\usepackage{hyperref}
\usepackage{fullpage}
\usepackage[utf8]{inputenc}


\begin{document}

\author{Sébastien Boisvert}
\title{Instruction Manual for the Ray \emph{de novo} genome assembler software}
\maketitle


Ray version 1.2.4

\vspace{2cm}

\href{http://denovoassembler.sf.net}{http://denovoassembler.sf.net}

\vspace{2cm}

Reference to cite: 
\vspace{2cm}

\noindent
Sébastien Boisvert, François Laviolette \& Jacques Corbeil.\\
Ray: simultaneous assembly of reads from a mix of high-throughput sequencing technologies.\\
\emph{Journal of Computational Biology} (Mary Ann Liebert, Inc. publishers, New York, U.S.A.).\\
November 2010, Volume 17, Issue 11, Pages 1519-1533.\\
doi:10.1089/cmb.2009.0238\\
\href{http://dx.doi.org/doi:10.1089/cmb.2009.0238}{http://dx.doi.org/doi:10.1089/cmb.2009.0238}

\newpage
\tableofcontents
\newpage

\section{Installation}

To install Ray, you need a C++ compiler, make and an MPI implementation compliant
with MPI standard 2.2.
These software are readily available in most GNU/Linux distributions.

\begin{table}{h}
\caption{Suggested software}
\begin{tabular}{ll}
\hline
Software & Name \\
\hline
C++ compiler & GNU g++ \\
MPI implementation & Open-MPI \\
\hline
\end{tabular}
\end{table}



\subsection{Compilation flags}

\begin{table}[h]
\caption{Compilation flags}
\begin{tabular}{ll}
\hline
Flag & Effet \\
\hline
HAVE\_ZLIB & enable support for .gz files \\
HAVE\_LIBBZ2 & enable support for .bz2 files \\
FORCE\_PACKING & enable packing of structures and classes\\
ASSERT & enable assertions in the code \\
\hline
\end{tabular}
\end{table}

\subsection{Compilation with the configure script}


\begin{verbatim}
./configure --prefix=$(pwd)/software/ray-x.y.z
make
make install
ls -l $(pwd)/software/ray-x.y.z/bin/Ray
\end{verbatim}

\subsection{Compilation with the makefile.alternate}

\begin{verbatim}
make -f makefile.alternate
ls -l code/Ray
\end{verbatim}

\section{Inputs}

The input files for Ray contain sequences. The files must be formatted in one of the supported formats.

\begin{table}[h]
\caption{File formats compatible with the Ray \emph{de novo} genome assembler software}
\begin{tabular}{lll}
\hline
Format & Obligatory extension \\
\hline
Fasta format & .fasta\\
Fasta format, compressed with GNU zip (gzip) & .fasta.gz \\
Fasta format, compressed with bzip2 & .fasta.bz2 \\
Fastq format & .fastq\\
Fastq format, compressed with GNU zip (gzip) & .fastq.gz \\
Fastq format, compressed with bzip2 & .fastq.bz2 \\
Standard flowgram format & .sff \\
\hline
\end{tabular}
\end{table}



\section{Parameters}

Ray assembles reads (paired or not) to produce an assembly.
Paired reads must be on opposite strands (forward \& reverse or reverse \& forward).

\subsection{$k$-mer length with -k}

Ray builds a distributed catalog of all occuring $k$-mers in the reads and their reverse-complement. $k$ must be greater or equal to 15 and lower or equal to 31. The $k$-mer length must be an odd number.

\subsection{Output prefix with -o}

Output files are named according to the prefix provided by the option -o.

\subsection{Single-end reads with -s}

\begin{verbatim}
-s <sequencesFile>
\end{verbatim}

\subsection{Paired-end reads and mate-pair reads with -p}

\begin{verbatim}
-p <leftSequencesFile> <rightSequencesFile> [ <averageFragmentLength> <standardDeviation> ]
\end{verbatim}

Example for paired-end reads (the ends of DNA fragments):

\begin{verbatim}
-p s_200_1.fastq s_200_2.fastq
\end{verbatim}

Example for mate-pair reads:

\begin{verbatim}
-p s_20000_1.fastq s_20000_2.fastq
\end{verbatim}

\subsection{Paired-end reads and mate-pair reads with -i (interleaved sequences)}

\begin{verbatim}
-i <sequencesFile [ <averageFragmentLength> <standardDeviation> ]
\end{verbatim}

In the interleaved file (example is for a fasta file):

\begin{verbatim}
>200_1_1234/1
ATCGATCGATCGACTCAGACACGTACG
>200_1_1234/2
ACTGACGACGTACGACGTCATGCAACT
...
\end{verbatim}



\section{Output}

\subsection{Contiguous sequences}

OutputPrefix.fasta contains contiguous sequences.

\subsection{Paired-end and mate-pair libraries}

OutputPrefix.LibraryLibraryNumber.txt contains the distribution of distances for paired-end and mate-pair libraries. One file per library.

\subsection{Coverage distribution}

OuputPrefix.CoverageDistribution.txt contains the $k$-mer coverage distribution.

\section{Example}

\subsection{Bacterial genome with paired-end and mate-pair short reads}

The command:

\begin{verbatim}
mpirun -np 32 ~/Ray/trunk/code/Ray \
-p /home/boiseb01/nuccore/Large-Ecoli/200_1.fastq \
   /home/boiseb01/nuccore/Large-Ecoli/200_2.fastq \
-p /home/boiseb01/nuccore/Large-Ecoli/1000_1.fastq \
   /home/boiseb01/nuccore/Large-Ecoli/1000_2.fastq \
-p /home/boiseb01/nuccore/Large-Ecoli/4000_1.fastq \
   /home/boiseb01/nuccore/Large-Ecoli/4000_2.fastq \
-p /home/boiseb01/nuccore/Large-Ecoli/10000_1.fastq \
   /home/boiseb01/nuccore/Large-Ecoli/10000_2.fastq \
-o BacterialGenome | tee RayLog
\end{verbatim}

\section{Virtual sequencer \& simulations}

The Ray package includes a simulator for paired reads.

\begin{verbatim}
N=6000000
readLength=50
errorRate=0.005
ref=~/nuccore/Ecoli-k12-mg1655.fasta

g++ code/simulatePairedReads.cpp -O3 -Wall -o Simulator 
./Simulator $ref $errorRate 200 20 $N $readLength L1_1.fasta L1_2.fasta
\end{verbatim}

\end{document}
